### ISSUE TITLE
addEventListener for "message" not relayed from browser extension

### ISSUE THREAD

**Issue by kasbah** [cypress-io/cypress#22391]
### Current behavior

When testing a page where a browser extension content script sends a message via `window.postMessage` and a page is listening via `window.addEventListener("message", ...)` Cypress does not pass on this message.

### Desired behavior

Cypress should pass on these events.

### Test code to reproduce

```
git clone https://github.com/kasbah/cypress-add-event-listener-repro
cd cypress-add-event-listener-repro
npm ci
npm run test
```

### Cypress Version

10.1.0

### Other

May be related to https://github.com/cypress-io/cypress/issues/1567

**Comment by lmiller1990** [comment]
Hi! 

I think there's a misunderstanding about how `launchOptions.extensions` works here, [docs](https://docs.cypress.io/api/plugins/browser-launch-api#Add-browser-extensions). This is for loading extensions **into the browser Cypress is running in**. I added a console.log into the content-script:

![image](https://user-images.githubusercontent.com/19196536/174928946-f7c035db-4a93-41a2-bf57-6d3379148401.png)

You can see this is logged immediately - this is executed as soon as the browser (which Cypress is running in) opens. If i then navigate to the spec:

![image](https://user-images.githubusercontent.com/19196536/174929044-f1b69bad-6059-4fd1-ba00-3b1e05546ddf.png)

We add the event listener - this is added to the `<iframe>` that your spec runs in. This is what is happening here:

1. extension is added to the main window
2. executes immediately
3. run spec
4. event added to the `<iframe>` which your application (localhost:3000) is rendered.

What are you trying to do? Are you looking to write E2E tests for your web extension? This is definitely something that you should be able to do with Cypress, but in a different fashion. Depending on what your extension is, you could host it on a server (say localhost:3000) and visit there using `cy.visit` - that should work as expected.

Check the [events catalog](https://docs.cypress.io/api/events/catalog-of-events#Cypress-Events) - maybe something like [this example](https://docs.cypress.io/api/events/catalog-of-events#Window-Before-Load) might be closer to what you want, eg:

```js
it('can listen', () => {
  cy.on('window:before:load', (win) => {
    win.addEventListener('...')
  })
})
```

If you can provide more insight into what you are trying to test, I can offer more suggestions.

**Comment by kasbah** [comment]
Hey, thanks for the detailed response. Indeed I am trying to test a browser extension. In this instance I am trying to test that a browser extension content-script is doing the right thing on a particular page. I'm struggling a bit with how I would do that by serving the extension up on localhost. Do you have any more links on that?

In case more details help: this is [the actual test](https://github.com/kitspace/web-extension/blob/4a804d5d49c7eda2ce9a01b81c301e8074cb6a0f/test/kitspace.spec.ts) I've been writing so far using Playwright. It visits [a page on our site](https://kitspace.org/boards/github.com/kitspace/ruler/) and clicks the "Mouser" button, which only does something useful with the extension loaded. We use Cypress extensively (on [kitspace-v2](https://github.com/kitspace/kitspace-v2)) so it'd be nice to stick with it for these tests.

**Comment by lmiller1990** [comment]
Right, I understand better now. When I suggested serving the extension locally, what I had in mind was that your extension is basically a mini web app (which it is, looks like a React app):

![image](https://user-images.githubusercontent.com/19196536/175200759-39aa675e-000f-4513-908c-9f80490528d1.png)

When I've made complex web extensions before, I've tested them by starting the dev server (eg `yarn start`) and my back-end API server (however you do it, for example `yarn backend`, navigated to it using `cy.visit` and testing it as if it's a isolated web app, using things like `cy.intercept` to assert API calls are made correctly. Now I better understand what you are doing here, I don't think this is really what you want.

I had a better look and we can load chrome extensions into the browser, the way you suggested - and since we change domain to match the iframe, I think your extension should be able to inject custom code into the app in the iframe, too.

My understanding here is you would like to test this in a true end-to-end fashion, with the actual extension packaged up and running in the browser. Navigating to a new tab is (currently) a limitation of Cypress. It is possible to change origin and follow a redirect in the same window using [cy.origin](https://docs.cypress.io/api/commands/origin). 

Having said that, maybe we can still write tests with Cypress. The goal, in general, that the extension: 

- injects some custom event listener on buttons
- when you click on a button, it makes an API call and pre-fills a cart
- **then** opens the page in a new tab? 

Is this correct? We should be able to do all of this, except asserting against the new tab (limitation of Cypress, at least today). I tried this but I'm enchanting something unexpected:

![Screen Shot 2022-06-23 at 2 58 51 pm](https://user-images.githubusercontent.com/19196536/175216714-a8c8e824-d31d-4332-98f8-f448d1dff8c8.png)

![Screen Shot 2022-06-23 at 3 00 10 pm](https://user-images.githubusercontent.com/19196536/175216764-995f684a-a8a1-48ed-a1ba-cf1f2e50d533.png)

I'd like to know why it isn't, at least, taking me to the correct tab (with the shopping cart). [Is this the right code base](https://github.com/kitspace/1clickBOM)? Can you point me to what exactly your chrome extension does in the background? I'd like to figure out what's going on here, since this should definitely be redirecting to the correct location, at least, based on my understanding.

**Comment by kasbah** [comment]
Hey, it is the right code base indeed. It can currently take 3 seconds to [register with the page](https://github.com/kitspace/1clickBOM/blob/12910e6dbfa46318b613ecfa8e0ee312b1f155bd/src/common/js/kitspace.js#L3-L6) which is why you aren't getting to the right page. You need to wait just over 3 seconds before clicking the button. 

I realise that the way this extension currently works has some problems and is not the best way to achieve the desired behavior.T his will be fixed in a re-write. I'ts part of why I'd like to have some e2e tests for the extension: to help with the re-write.  

Unfortunately, in addition to all the e2e frameworks I've tried not being that well set up to test extensions yet, I've hit another road block: in interacting with the external site during tests (mouser.com) we sometimes get flagged as a bot and get blocked. puppeteer-extra-plugin-stealth seems to be the best low-effort way to try and avoid this, but that means I need to use puppeteer I guess. I might have to give up on the idea of using e2e testing for this.

**Comment by lmiller1990** [comment]
Thanks for clarifying. I'll try and take another look at this, sometime this week. It seems like there's a few other blockers? I'm not sure I can prioritize this one right now, but I'm definitely interested in the general "testing extensions with Cypress" problem, so I'll try and loop back around to this later this week. Let me know if you come up with anything in the meantime.
